### PR TITLE
Completed move_zeroes algorithm

### PR DESCRIPTION
**Pull request by maxhealy01**
## Summary
This is a solution to the Move Zeroes problem, linked below.

## Thought Process
This solution is written to alter an array of numbers in-place, which explains some of its quirks. I remove elements from the array during a for-loop, and decrement the index accordingly, which might cause some confusion but makes sense to me.

The only extra space used was for the zeroCount variable.

## Issue
https://github.com/maxhealy01/blind-75/issues/3

## Checklist
- [x] Contains JSDoc comments
- [x] Contains tests
- [x] Completed self review

### CODE REVIEW

File: week_1/Array/move_zeroes.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.
+
+// Note that you must do this in-place without making a copy of the array.
```
Comment (tannaurus) [inline week_1/Array/move_zeroes.js:3]: This looks like a left over from the test instructions, we should rephrase this to be more of a warning than a footnote. Generally, at least in JavaScript, mutating params can be a _very_ bad thing and lead to _very_ strange behavior. It can be an effective memory savor when you are expecting it, though 😄 

For context, let's look at how Rust handles mutation. This function's signature would look something like this:
```rust
fn moveZeroes(&mut nums)
```

The `&mut` here tells the person who is about to use this function that mutation is more than likely going to happen (otherwise, it's just bad Rust code 😄 )

If you wanted to tell the caller that you _weren't_ going to mutate anything, the function signature would look like this:
```rust
fn moveZeros(&nums
```

The omission of `mut` means any code in the function can't mutate the param it was given! If you were to try, your code won't compile 🚀 

All of this makes it super clear when you plan on mutating and when you aren't. In languages like JavaScript where you can't express this that easily, you want to take the time to explicitly call out when you are mutating. Don't be afraid to be loud about it 😄 it's typically common, in a good code base, to assume things aren't mutating; with a some exceptions that come to mind. 

_note, none of that is real Rust code, trimmed off a couple things that weren't applicable 😄  but it's close!_

File: week_1/Array/move_zeroes/move_zeroes.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+// This operation occurs in place, meaning that the array is mutated. Be careful!
+/**
+ * @param {number[]} nums
+ * @return {void} Do not return anything, modify nums in-place instead.
+ */
+const moveZeroes = (nums) => {
+	let zeroCount = 0;
+	for (i = 0; i < nums.length; i++) {
+		if (nums[i] === 0) {
```
Comment (tannaurus) [inline week_1/Array/move_zeroes/move_zeroes.js:11]: I think this solution would be much easier to read if it didn't attempt to mutate each index in place, but rather the entire array.

```js
let zeroCount = 0;
let filteredNumbers = nums.filter(num => num === 0);
// combine the filter numbers with a new array that full of zeros
nums = [...filteredNumbers, ...Array(zeroCount).fill(0)]
```

It would be interesting to see a comparison of performance between the two solutions here. I'd wager your original solution is more performant, but it's a bit difficult to comprehend.

File: week_1/Array/move_zeroes/move_zeroes.test.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+const moveZeroes = require("./move_zeroes");
+
+test("move_zeroes moves zeroes to the end of an array of numbers while maintaining the relative order of the remaining elements", () => {
+	let nums = [0, 10, 0, 2, 3, 4, 0, 5, 6, 7];
```
Comment (tannaurus) [inline week_1/Array/move_zeroes/move_zeroes.test.js:4]: can we add a test that asserts nothing is changed when there are no zeros, or when the zeros are already on the end?

Comment (maxhealy01) [reply week_1/Array/move_zeroes/move_zeroes.test.js]: Done

File: week_1/Array/move_zeroes.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * @param {number[]} nums
+ * @return {void} Do not return anything, modify nums in-place instead.
+ */
+const moveZeroes = (nums) => {
```
Comment (tannaurus) [inline week_1/Array/move_zeroes.js:8]: looks like there is two duplicate files here?

File: week_1/Array/move_zeroes/move_zeroes.js (hunk 2 above)
Comment (tannaurus) [reply week_1/Array/move_zeroes/move_zeroes.js:11]: I see now that a requirement for this issue was to do this "in place".

I'm beginning to become a bit concerned that these requirements are encouraging bad practices. These type of requirements are less problematic for languages with stronger type systems, but very scary to see in JavaScript (zero type safety).

File: week_1/Array/move_zeroes.js (hunk 4 above)
Comment (maxhealy01) [reply week_1/Array/move_zeroes.js]: I got rid of the one in the Array folder.

**Comment by tannaurus** [comment]
Looks great 👍
